Add more openat() style variants for existing stuff

DaanDeMeyer opened the pull request:
I'm still adding a lot more wrappers but feel free to review the stuff that's already in here.

(I still need to update the big documentation blob in chase_symlinks_at())

DaanDeMeyer commented:
Should be ready for review now. My goal here is to allow constructing directory hierarchies in tests in a temporary directory without having to path_join() all the time by just using the temporary directory file descriptor as a directory fd and use these openat() helpers to specify relative paths to that fd.

poettering commented on src/basic/tmpfile-util.c:
```diff
@@ -36,12 +36,12 @@ int fopen_temporary(const char *path, FILE **ret_f, char **ret_temp_path) {
 
-        fd = mkostemp_safe(t);
+        fd = openat(dir_fd, t, O_CLOEXEC|O_NOCTTY|O_RDWR|O_CREAT|O_EXCL, 0600);
         if (fd < 0)
```
the error path below needs a change from unlink() to unlinkat() too

poettering commented on src/basic/path-util.c:
```diff
@@ -221,6 +221,33 @@ int path_make_relative_parent(const char *from_child, const char *to, char **ret
+                        return r;
+        }
+        if (!c)
+                return -ENOMEM;
```
this error path belongs to the if block, it's pointless on the else block.

poettering commented on src/basic/path-util.c:
```diff
@@ -221,6 +221,33 @@ int path_make_relative_parent(const char *from_child, const char *to, char **ret
+        assert(p);
+        assert(ret);
+
+        if (!path_is_absolute(p))
```
if you invert the test you could switch the if/else blocks an save ONE! character!

poettering commented on src/basic/chase-symlinks.h:
```diff
@@ -23,7 +23,7 @@ bool unsafe_transition(const struct stat *a, const struct stat *b);
 /* How many iterations to execute before returning -ELOOP */
 #define CHASE_SYMLINKS_MAX 32
 
-int chase_symlinks(const char *path_with_prefix, const char *root, ChaseSymlinksFlags chase_flags, char **ret_path, int *ret_fd);
```
the function carried this name because unless you specify CHASE_PREFIX_ROOT it must be the path with the root prefix already prepended, which is noteworthy, because I guess people in many contexts would not expect that.

poettering commented on src/basic/fs-util.c:
```diff
@@ -195,13 +195,13 @@ int readlink_and_make_absolute(const char *p, char **r) {
+                                                                 * to change mode/owner on the same file */
         if (fd < 0)
                 return -errno;
 
```
hmm, if path is specified as NULL, why not shortcut this to fchmod_and_chown() and pass dir_fd? sounds more versatile that way.

poettering commented on src/basic/chase-symlinks.c:
```diff
@@ -57,22 +57,17 @@ static int log_autofs_mount_point(int fd, const char *path, ChaseSymlinksFlags f
-
-        _cleanup_free_ char *buffer = NULL, *done = NULL, *root = NULL;
-        _cleanup_close_ int fd = -1;
+int chase_symlinks_at(int dir_fd, const char *path, ChaseSymlinksFlags flags, char **ret_path, int *ret_fd) {
```
i'd really break the arguments line here into multiple lines. Once we have 3 or so arguments it's just hard to read.

poettering replied:
what is "dir_fd"? is that the dir we take the path relative to? i.e. the *root* dir?

poettering commented on src/basic/chase-symlinks.c:
```diff
@@ -400,6 +394,88 @@ int chase_symlinks(
+        /* chase_symlinks_at() disallows empty paths, so if the "path" is identical to "root", let's
+         * explicitly specify that we want to open the root directory by adding ".". */
+        if (isempty(path))
+                path = ".";
```
wouldn't it make more sense to pass `path` as NULL in that case and let chase_symlinks_at() figure it out

poettering commented on src/basic/chase-symlinks.c:
```diff
@@ -113,69 +114,74 @@ int chase_symlinks(
-                if (r < 0)
-                        return r;
-        }
+        rootfd = fd_reopen(MAX(dir_fd, -1), O_CLOEXEC|O_DIRECTORY|O_PATH);
```
hmmm? `MAX(dir_fd, -1)`?

DaanDeMeyer replied:
To ignore `AT_FDCWD` basically, but I now see that this is rather unreadable and brittle, since it depends on `AT_FDCWD` being smaller than -1, so let me rework this

poettering commented on src/basic/chase-symlinks.c:
```diff
@@ -57,22 +57,17 @@ static int log_autofs_mount_point(int fd, const char *path, ChaseSymlinksFlags f
         int r;
 
         assert(path);
+        assert(!FLAGS_SET(flags, CHASE_PREFIX_ROOT));
```
we probably should refuse root_fd here if negative and != AT_FDCWD, no?

poettering replied:
or is -1 supposed to be "relative to `/`"?

poettering commented on src/basic/chase-symlinks.c:
```diff
@@ -113,69 +114,74 @@ int chase_symlinks(
-                original_root = NULL;
-
-        if (!original_root && !ret_path && !(flags & (CHASE_NONEXISTENT|CHASE_NO_AUTOFS|CHASE_SAFE|CHASE_STEP)) && ret_fd) {
+        if (dir_fd < 0 && !ret_path && !(flags & (CHASE_NONEXISTENT|CHASE_NO_AUTOFS|CHASE_SAFE|CHASE_STEP)) && ret_fd) {
```
so what's the goal here? you test dir_fd < 0, but then pass that to openat(), which will cause an error if you pass anything but AT_FDCWD.. but if you want to only do that on AT_FDCWD, then maybe write this as `dir_fd == AT_FDCWD`? and in that case you don't need openat() anyway, but can stick to open()

DaanDeMeyer replied:
I was trying to follow `openat()` behavior here. There's the following snippets in its man page:

> If the pathname given in pathname is absolute, then dirfd is ignored.

> If  the  pathname  given in pathname is relative, and dirfd is not a valid file descriptor, an error (EBADF) results. (Specifying an invalid file descriptor number in dirfd can be used as a means to ensure that pathname is absolute.)

So -1 is actually a valid value for `dir_fd`, as long as the path is absolute. But thinking about it, there's no real reason to support that in `chase_symlinks_at()`, since users can easily just pass `AT_FDCWD` with an absolute path to get exactly the same behavior, so we can enforce that `dir_fd` is either positive or `AT_FDCWD`.

poettering commented on src/basic/mkdir.c:
```diff
@@ -31,27 +32,27 @@ int mkdir_safe_internal(
                 _cleanup_free_ char *p = NULL;
 
-                r = chase_symlinks(path, NULL, CHASE_NONEXISTENT, &p, NULL);
+                r = chase_symlinks_at(dir_fd, path, CHASE_NONEXISTENT, &p, NULL);
```
hmm, so this looks wrong. the second argument to chase_symlinks() was the root filesystem we take absolute symlinks relative to. But in your new code you actually take them relative to dir_fd?

DaanDeMeyer replied:
So it's kinda complicated. The best way to solve this would be to have chase_symlinks_at() take both a root directory fd and a separate directory fd. But to get that to work, we need a way to transform the path relative to the directory fd to a path relative to the root directory fd. The only way I know of to do that is via /proc. If that's viable, it should be possible to add chase_symlinks_at_full() that takes a root directory fd, a directory fd and a path relative to the directory fd, transform the path to be relative to the root directory fd via /proc and then call chase_symlinks_at() with the transformed path and the root directory fd. 

I'm not sure if the current behavior makes sense either though. If we look at the usages of `mkdir_parents_safe()` and `mkdir_p_safe()` in nspawn-mount.c, they pass a root directory as the prefix, but the symlinks are still resolved against the host's root directory, that doesn't seem correct to me.

poettering commented:
i'd prefer that if we'd add these and have tests and at least one actual user of each func...

DaanDeMeyer commented:
Reworked `chase_symlinks_at()` quite a bit. It now behaves a lot more like openat2() and it's `RESOLVE_IN_ROOT` flag. Specifically, we only use the provided directory fd as the root directory to resolve against if the new `CHASE_AT_RESOLVE_IN_ROOT` is set. Otherwise we resolve against the root directory.

I also reworked `test_copy_tree_replace_dirs()` to use the new at helpers and mkdtemp_open(). Overall idea is that every test that uses a temporary directory uses at helpers exclusively instead of constantly joining paths.

bluca commented on src/basic/tmpfile-util.c:
```diff
@@ -358,3 +358,23 @@ int mkdtemp_malloc(const char *template, char **ret) {
+        fd = open(p, O_DIRECTORY|O_CLOEXEC|flags);
+        if (fd < 0) {
+                (void) rmdir(p);
+                return -errno;
```
if rmdir fails errno will no longer be what `open` returned, no?

bluca commented on src/basic/fs-util.c:
```diff
@@ -195,17 +195,17 @@ int readlink_and_make_absolute(const char *p, char **r) {
-int chmod_and_chown(const char *path, mode_t mode, uid_t uid, gid_t gid) {
+int chmod_and_chown_at(int dir_fd, const char *path, mode_t mode, uid_t uid, gid_t gid) {
         _cleanup_close_ int fd = -1;
 
```
assert on dird_fd

DaanDeMeyer replied:
`openat()` like calls generally don't put any requirements on `dir_fd`, it can be negative, `AT_FDCWD` or positive and all of those mean different things, so I don't think we should assert here.

bluca commented on src/basic/mkdir.c:
```diff
@@ -31,27 +32,27 @@ int mkdir_safe_internal(
         assert(mode != MODE_INVALID);
         assert(_mkdirat && _mkdirat != mkdirat);
 
```
assert on dir_fd

mrc0mmand commented:
(Don't mind the failed CentOS CI jobs - there was an issue with the provisioner and the failed jobs will get restarted once the queue frees up a bit)

poettering commented on src/shared/rm-rf.h:
```diff
@@ -51,3 +52,19 @@ static inline char *rm_rf_subvolume_and_free(char *p) {
+                return safe_close(fd);
+
+        safe_close(fd);
+        (void) rm_rf(p, REMOVE_ROOT|REMOVE_PHYSICAL|REMOVE_MISSING_OK|REMOVE_CHMOD);
```
hmpf, i don't really like this I must say fd_get_path() is entirely unreliable in particular for stuff that has already been deleted (as the path it reports will then be suffixed with `(removed)`, and thus is entirely ambigious). 

Sorry, but this cannot work I fear.

fd_get_path() is OK for some things (informational stuff for example), but I think using it for destructive file operations is a big problem.

This goes particularly bad if the fd passed refers to something that is not actually a thing in the file system (sockets, eventfd, …).

I#d really avoid this.

DaanDeMeyer replied:
Dropped this stuff again

poettering reviewed (commented):
sorry I have been a bit slow on reviewing this.

DaanDeMeyer commented on src/basic/chase-symlinks.c:
```diff
@@ -113,97 +124,72 @@ int chase_symlinks(
+                return -errno;
 
-        fd = open(empty_to_root(root), O_CLOEXEC|O_DIRECTORY|O_PATH);
+        if (FLAGS_SET(flags, CHASE_AT_RESOLVE_IN_ROOT) || !path_is_absolute(path))
```
It is, but `need_absolute` is a bad description to use here so I went for the full condition.

DaanDeMeyer commented on src/basic/path-util.c:
```diff
@@ -221,6 +221,33 @@ int path_make_relative_parent(const char *from_child, const char *to, char **ret
+
+        assert(p);
+        assert(ret);
+
```
This function was actually unused, so I just dropped it

DaanDeMeyer commented on src/basic/chase-symlinks.c:
```diff
@@ -401,6 +389,88 @@ int chase_symlinks(
+        }
+
+        if (root) {
+                fd = open(root, O_CLOEXEC|O_DIRECTORY|O_PATH);
```
This causes a test failure and we didn't do this before earlier so since it changes behavior let's leave it for a separate PR.

DaanDeMeyer commented:
See #25338
